### PR TITLE
refactor(achievements): implement event bus and annotation-based auto-registration (#357)

### PR DESCRIPTION
**Pull request by Teo418**
### ✅ PR Checklist
- [x] My PR title follows the format: `type(scope): description (#issue)` (scope and #issue optional)
- [x] I used the correct `type`: feat, fix, refactor, docs, test, chore
- [x] I reviewed my code and removed unnecessary debug logs or comments.
- [x] I tested my changes and verified they work as expected.
- [x] I ran the project to confirm it compiles and runs correctly.
- [x] I read the [Code of Conduct](https://github.com/jvondermarck/dinosaur-exploder/blob/main/CODE_OF_CONDUCT.md) and the [Contribution Guidelines](https://github.com/jvondermarck/dinosaur-exploder/blob/main/CONTRIBUTING.md).

### 📝 What does this PR do?

Part of #357

This PR refactors the achievement system to eliminate the linear growth problem reported in the issue. Two complementary solutions were implemented:

**Event Bus (`dispatch()`):**
- Replaces all individual `notifyDinosaurKilled()`, `notifyScoreChanged()`, `notifyCoinCollected()` and `notifyBossDefeated()` methods with a single `dispatch(AchievementEvent)` entry point
- Adding a new event type no longer requires modifying `AchievementManager`
- `CollisionHandler` and `ScoreComponent` were updated to use `dispatch()` directly

**Annotation-based auto-registration (`@RegisterAchievement`):**
- Replaces the hardcoded `AchievementCatalog.defaults()` with a classpath scanner that discovers annotated `Achievement` subclasses automatically at startup
- Uses pure Java reflection — no external dependencies added
- Works in all deployment layouts: exploded directory (tests/IDE) and fat JAR
- Adding a new achievement now only requires creating the class and annotating it — no other file needs to change

**Files changed:**
- New: `RegisterAchievement.java`, `RegisterAchievements.java` (repeatable annotation container)
- Modified: `AchievementManager.java`, `AchievementCatalog.java`, `AchievementEvent.java`, `AchievementEventType.java`
- Modified: all 5 achievement subclasses (annotations added)
- Modified: `CollisionHandler.java`, `ScoreComponent.java` (migrated to `dispatch()`)
- Modified: `AchievementTest.java` (updated and expanded)

### 🔗 Related Issue
- [x] This PR fixes/closes: #357

### 💬 Extra Notes (Optional)
This was my first contribution to an open-source project. Really enjoyed the process and the codebase. Happy to make any adjustments based on review feedback!

### CODE REVIEW

**Review by Teo418 — COMMENTED** [review]
This hotspot is triggered by the JarFile usage in the classpath scanner 
(collectFromJar). The JAR being opened is always the application's own 
JAR retrieved via ClassLoader.getResources(), so there is no user-controlled 
input involved. Safe to mark as acknowledged.

File: .vscode/settings.json
Comment (jvondermarck) [inline .vscode/settings.json:1]: Please don't push this file, maybe exclude it from the `.gitignore`

File: src/main/java/com/dinosaur/dinosaurexploder/achievements/AchievementManager.java [hunk 1]
```diff
@@ -151,4 +160,139 @@ public void saveAchievement(List<Achievement> listToSave) {
...
+  // Finds annotated classes, sorts them for deterministic order, and builds the catalog.
+  @SuppressWarnings("unchecked")
+  private static AchievementCatalog buildCatalogFromAnnotations() {
+    List<Class<? extends Achievement>> classes = findAnnotatedClasses();
+    classes.sort(Comparator.comparing(Class::getSimpleName));
+    return AchievementCatalog.of(buildFactories(classes).toArray(Supplier[]::new));
+  }
+
+  // Converts a sorted list of annotated classes into one factory per @RegisterAchievement entry.
+  private static List<Supplier<Achievement>> buildFactories(
```
Comment (jvondermarck) [inline src/main/java/com/dinosaur/dinosaurexploder/achievements/AchievementManager.java]: I'm not very good in Java, but all of this code from the line 144 is a bit complicated to understand, and mainly to maintain in the future. Like why we need to do stuff with `jar` and `package` ? 

Obviously now code looks better with this `@RegisterAchievement(target = 500, reward = 200)` but the setup is looking very difficult to understand in my opinion. Like my question is maybe there is a way to make it easier, maybe it's an over-engineering implementation?

Share me your thoughts on this code and find a solution together.

Nevertheless, thanks so much for this work! Let's first discuss together :)

**Review by jvondermarck — COMMENTED** [review]
I have some thoughts about the implementation, so let's see together before merging or whatever

File: .vscode/settings.json
Comment (Teo418) [reply .vscode/settings.json:1]: Hey! Yes totally my bad, I'll add .vscode/ to the .gitignore right away. Should I just push that fix directly to the same branch so the PR updates automatically, or do you prefer I wait until we've aligned on the other points first?

File: src/main/java/com/dinosaur/dinosaurexploder/achievements/AchievementManager.java (hunk 1 above)
Comment (Teo418) [reply src/main/java/com/dinosaur/dinosaurexploder/achievements/AchievementManager.java]: You're completely right, and honestly I think you put it better than I did. The code from line 144 exists to "automatically discover" achievement classes without knowing their names in advance, which is why it needs to deal with JARs, ClassLoaders and file system paths — it's literally scanning the compiled output looking for classes with the @RegisterAchievement annotation. That's powerful but way too complex for what this project actually needs.

The simpler approach would be to keep the @RegisterAchievement annotations on each class (since you liked that part) but replace the entire scanner with a plain static list inside AchievementManager:

private static final List<Class<? extends Achievement>> ACHIEVEMENT_CLASSES = List.of(
    BossDefeatAchievement.class,
    CoinCollectionAchievement.class,
    KillCountAchievement.class,
    ScoreAchievement.class,
    SurvivalTimeAchievement.class);

Then the registration logic just reads the annotations from that list directly. No ClassLoader, no JARs, no file system scanning. The whole auto-registration section would go from ~80 lines to around 10, and anyone reading it would understand it immediately.

The only difference is that adding a new achievement would require editing 2 files instead of 1 (create the class + add one line to the list), but I think that's a very reasonable trade-off for code that's actually maintainable.

What do you think? If this direction works for you I can update the PR with this simplified version.

Comment (jvondermarck) [reply src/main/java/com/dinosaur/dinosaurexploder/achievements/AchievementManager.java]: Thanks so much for your answer @Teo418 !

> That's powerful but way too complex for what this project actually needs.

Exactly, it's powerful, it's not bad, maybe for a production app with millions of users and a huge codebase, why not. But well, in our case, the game is small, we need to make it easier for people with not much experience on Java could work on the codebase without too much issues (even if now I guess people use ChatGPT for most things aha).

> Then the registration logic just reads the annotations from that list directly. No ClassLoader, no JARs, no file system scanning. The whole auto-registration section would go from ~80 lines to around 10, and anyone reading it would understand it immediately.

Your new solution looks 100% perfect in my opinion yes, and don't worry if we edit two files it's not a problem as it's not the quantity BUT the quality. And let's be honest, it takes 2 seconds for the dev to add the new file in the list, so that's something we can call it "generic", as easy to add new things easily without breaking the code and being quick.

> What do you think? If this direction works for you I can update the PR with this simplified version.

So yes let's do this then, you can update the PR with this version, and you can also update the test cases so that we have some code coverage 😎

---

Thanks a lot again for your quick answer and your solution ! That is what is needed from an engineer ! ❤️

File: .vscode/settings.json
Comment (jvondermarck) [reply .vscode/settings.json:1]: Yeeees no problem, you can push directly in the same branch, so in this PR to update the `.gitignore` as it is just one file and one line to update, so no problem :) You can ! 

Thanks a lot

Comment (jvondermarck) [reply .vscode/settings.json:1]: My bad to me, it's weird the `.vscode` folder is not excluded yet ahha, it's been few times I see people pushing this, but forgot to have a look why and check the `.gitignore` ahha

Comment (Teo418) [reply .vscode/settings.json:1]: Haha no worries, just pushed the .gitignore fix to the same branch, the PR should be updated now!

File: src/main/java/com/dinosaur/dinosaurexploder/achievements/AchievementManager.java (hunk 1 above)
Comment (Teo418) [reply src/main/java/com/dinosaur/dinosaurexploder/achievements/AchievementManager.java]: Thank you so much for the kind words, really appreciated! 😄

Completely agree with your take — keeping it simple and accessible for contributors is way more valuable than a clever solution that nobody wants to touch. I'll update the PR with the simplified version (static list + annotations, no scanner) and update the test cases to cover the new design. I'll let you know once it's ready for review!

Comment (Teo418) [reply src/main/java/com/dinosaur/dinosaurexploder/achievements/AchievementManager.java]: Hey! Just pushed the updated version to the same branch, the PR is ready for review whenever you want 😊

Quick summary of what changed: replaced the entire classpath scanning logic with a simple static list of classes inside AchievementManager, so the auto-registration section went from ~80 lines down to about 10. Also added 5 new test cases to cover all the event types (score, coins, boss defeat, survival time) and the null case when there are no active achievements. Let me know if you have any questions!

**Review by jvondermarck — APPROVED** [review]
Awesome now,thanks a lot for your contribution !! @Teo418 
Please "Star" the project, it brings support for more vision of the project :)
